FlowingCode/FontAwesomeIronIconset
# feat(generator): support up to 10921 constants

As discussed in https://github.com/FlowingCode/FontAwesomeIronIconset/issues/65#issuecomment-1450740812, enum classes are limited to 4103, even with newer compilers, because the enum fields must be assigned in the static initializer (per [putstatic specs](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-6.html#jvms-6.5.putstatic): "if the field is final, it must be declared in the current class, and **the instruction must occur in the <clinit> method of the current class**." --emphasis mine).

This PR enables an alternate compilation method (via a maven profile `-Pbigenum`) that generates size-optimized bytecode for the initialization of enumerated classes and allows up to 10921 constants per enum. (The add-on available in Vaadin directory is not affected because it has less than 4103 and can be compiled through the usual mechanism.)

## &lt;clinit>

The initialization in `<clinit>` is reduced to 6 bytes per constant, i.e. `CONSTANT = $init();`

```
0: invokestatic  Method $init:(I)Lcom/example/MyEnum;
3: putstatic     Field CONSTANT:Lcom/example/MyEnum;
6: ...
```

instead of the usual 15-16 bytes of  `CONSTANT = new Enum(name, ordinal);`

```	   
0: new           class com/example/application/MyEnum
3: dup
4: ldc/ldc_w     "name"
7: sipush        12345
10: invokespecial Method "<init>":(Ljava/lang/String;I)V
13: putstatic    Field CONSTANT:Lcom/example/application/MyEnum;
16: ...
```

In both cases, the static initializer ends with 7 bytes that call the synthetic method `$values()` and store its result into `$VALUES`:
```
65529 : invokestatic Method $values:()[Lcom/example/MyEnum;
65532 : putstatic Field $VALUES:[Lcom/example/MyEnum;
65535 : return
```

The approach includes several optimizations:

- 4 bytes of `new`+`dup` are claimed by replacing `new`+`dup`+`invokespecial` with `invokestatic` of an `$init` method that leaves the new instance in the stack. 
- 3 bytes of `sipush` are claimed by keeping the ordinal as a static field (incremented by `$init`), in order to avoid pushing a constant into the stack before each call.
- 3 bytes of `ldc_w` (2 bytes of `ldc`) can be claimed by computing the name from the ordinal in `$init`, so that the method takes no arguments.


The proposed approach use the following synthetic members:
```
private static transient int _ordinal;
private static String $init() { return new MyEnum($name(_ordinal), _ordinal++); }
private static String $name(int ordinal) { /*return name from ordinal*/ }
```

### $name

The `$name` method maps an ordinal to the corresponding name:

``` 
switch (ordinal) {
  case 0 : return "CONST0";
  case 1 : return "CONST1";
  //...
  default: throw new AssertionError();
}
```

This code translates to bytecode as:
```
 0: getstatic     Field _ordinal:I
 3: tableswitch   { // 0 to N
	0: 24
    1: 28
    // ... 
    default: 
   }
24: ldc_w         String "CONST0"
27: areturn
28: ldc_w         String "CONST1"
31: areturn
28: ... 
```

The approach requires 8 bytes per entry (tableswitch entry + ldc_w + areturn) plus 28 bytes of overhead (getstatic, tableswitch header, 4 bytes for the default case and 8 bytes for `throw new`). Thus it allows at most 8188 cases per method (a little bit more if `ldc` can be used instead of `ldc_w`, although it isn't a game changer and ASM makes hard to know which opcode is used).

In order to allow more constants, we need to split this method so that each chunk is below 64KB:
```
  private static String $name() {
    switch (ordinal) {
      case 0: return "CONST0";
      case 1: return "CONST1";
	  //...
      default: return $name2();
    }
  }
  
  private static String $name2() {
    switch (ordinal) {
      case 8188: return "CONST8189";
      case 8189: return "CONST8190";
	  //...
      default: throw new AssertionError();
    }
  }
```
  
### $values

The naive implementation of `$values()` allows up to 8207 constants: 

```
  private static C[] $values() {    
    C[] values = new C[N];
    values[0] = CONST0;
    values[1] = CONST1;
    //...
    return values;
  }
```

Which translates to bytecode as:
```	
         0: sipush        <array size>
         3: anewarray     class com/example/MyEnum

         6: dup
         7: sipush        index
        10: getstatic     Field CONST1:Lcom/flowingcode/vaadin/addons/fontawesome/FontAwesome$Solid;
        13: aastore

        14: ...
```
        
The bytecode has an overhead of 7 bytes, plus 8 bytes per constant.
(The first constants use iconst_i/bipush instead of sipush, which saves 134 bytes)

In order to split this method, we add a second method `$values_hi(C[])` that continues the assignment:

```
 private static C[] $values() {
    C[] values = new C[N];
    values[0] = CONST0;
    values[1] = CONST1;
    //...
    $values_hi(values); 
    return values;
  }

  private static void $values_hi(C[] values) {    
    values[8207] = CONST8207;
    values[8208] = CONST8208;
    //...
    return;
  }
```

The bytecode needed for calling `$values2()` has just 4 bytes, thus `$values()` can still accommodate 8207 assignments.
`$values_hi()` has an overhead of 2 bytes, plus 8 bytes per constant (fixed) and can accommodate 8191 assignments (which is enough for the 10921 constants allowed by the implementation of `<clinit>`).

## Tests

Several tests are added for validating correctness of each enum type:
- check that `Class.getEnumConstants` is not null (it can return null "when users concoct enum-like classes that don't comply with the enum spec")
- check that we have the same number of constants and constant fields of the enum type
- check that every constant in values correspond to a static final field of the enum type
- check that all the constants were returned by `values()`
- check that there are no gaps in the ordinal values

## Implementation

`IconsetEnumCompiler` has three phases: 
- rewrite (count how many constants there are, and optionally modify the source code)
- compile (compile by using the JDK compiler)
- postcompile (implement the synthetic methods)

The behavior depends on the amount of constants and whether postprocessing is allowed.

If the `bigenum` profile is enabled:
- The main `pom.xml` sets `maven.main.skip` in order to bypass compilation of main sources by `maven-compiler-plugin`
- If any enumeration exceeds 10921 constants, then fail.
- If any enumeration exceeds 4103 constants, then remove all the declared constants (so that compilation succeeds) and enable the postcompile phase.
- Run the compile phase.
- Run the postcompile phase (which adds back the constants using ASM).

If the `bigenum` profile is disabled:
- If any enumeration exceeds 4103 constants, then fail.
- Otherwise, skip compile and postcompile phases.
- Compilation will be performed by `maven-compiler-plugin`

## Other notes

If big enums are used, running the standalone demo with Jetty requires skipping the compilation phase (i.e. `mvn jetty:run -Dmaven.main.skip`). Otherwise maven will try to recompile the sources before starting jetty, and it will fail because it will use the standard compiler.
